EIPS/eip-7851.md

```diff
@@ -0,0 +1,247 @@
+assert error == b"insufficient gas"
+assert gas_left == 0
+
+# EOA without delegated code
+caller = "0x4567"
+error, gas_left = precompile.execute(caller, state_db, gas=10000)
+assert error == b"invalid delegated code prefix"
+assert gas_left == 0
+
+# Small contract code
+caller = "0x89ab"
+state_db.set_code(caller, bytes.fromhex("00")) # a fake contract
+error, gas_left = precompile.execute(caller, state_db, gas=10000)
+assert error == b"invalid delegated code prefix"
+assert gas_left == 0
+```
+
+## Reference Implementation
+
+```python
+class PrecompiledContract:
+    DELEGATED_CODE_PREFIX = bytes.fromhex("ef0100")  # EIP-7702 prefix
+    GAS_COST = 5000  # PRECOMPILE_GAS_COST
+
+    def execute(self, caller, state_db, gas, static=False):
+        """
+        Toggle EOA's private key authorization between active/deactivated states.
+
+        Parameters:
+        - caller: The address calling the contract
+        - state_db: The state database
+        - gas: Gas provided for execution
+        - static: Whether called in read-only context
+
+        Returns:
+        - Tuple of (result, gas_left)
+          result: error bytes on failure, empty bytes on success
+          gas_left: remaining gas, 0 on error
+        """
+        # Check gas
+        if gas < self.GAS_COST:
+            return b"insufficient gas", 0
+
+        # Check static call
+        if static:
+            return b"cannot call in static context", 0
+
+        # Get and validate caller's code
+        code = state_db.get_code(caller)
+        if not code.startswith(self.DELEGATED_CODE_PREFIX):
+            return b"invalid delegated code prefix", 0
+
+        # Update code based on length
+        if len(code) == 23:  # Active state
+            state_db.set_code(caller, code + b"\x00")  # Deactivate
+        elif len(code) == 24:  # Deactivated state
+            state_db.set_code(caller, code[:-1])  # Activate
+        else: # Although this is not possible, it's added for completeness
+            return b"invalid code length", 0
+
+        return b"", gas - self.GAS_COST
+
+class StateDB:
+    """Simplified state database, omitting other account fields"""
+    def __init__(self):
+        self.accounts = {}
+
+    def get_code(self, addr):
+        return self.accounts.get(addr, {}).get("code", b"")
+
+    def set_code(self, addr, value):
+        if addr not in self.accounts:
+            self.accounts[addr] = {}
+        self.accounts[addr]["code"] = value
+```
+
+## Security Considerations
+
+### Additional checks during transaction validation
+
+The deactivation status is determined by checking the length of the delegated code. This check introduces an additional storage read and check comparable to account state read and checks, such as nonce validation. It is integrated into the transaction validity check of the transaction pool and consensus rules, ensuring that invalid transactions are filtered out before propagation and execution.
```

Comment:
> This check introduces an additional storage read

Since we read the account anyway for the nonce, I don't think this requires an additional read, so there's no additional workload for the txpool.

Reply:
There is an extra storage read, because the code itself is not directly stored in the account: only the code hash is stored in the account. So an extra storage read is necessary to get the actual code of the account (and not the hash).